pkg/plugins/runtime/k8s/controllers/cni_taint_controller.go

```diff
@@ -0,0 +1,188 @@
+	"golang.org/x/exp/slices"
+	kube_core "k8s.io/api/core/v1"
+	kube_apierrs "k8s.io/apimachinery/pkg/api/errors"
+	kube_types "k8s.io/apimachinery/pkg/types"
+	kube_ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	kube_client "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	kube_handler "sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	kube_reconcile "sigs.k8s.io/controller-runtime/pkg/reconcile"
+	kube_source "sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	nodeReadinessTaintKey = "NodeReadiness"
+	nodeIndexField        = "spec.nodeName"
+	cniAppLabel           = "app"
+	cniPodNamespace       = "kube-system"
+)
+
+type CniNodeTaintReconciler struct {
+	kube_client.Client
+	Log logr.Logger
+
+	CniApp string
+}
+
+func (r *CniNodeTaintReconciler) Reconcile(ctx context.Context, req kube_ctrl.Request) (kube_ctrl.Result, error) {
+	log := r.Log.WithValues("node", req.NamespacedName)
+
+	// Fetch the Node instance
+	node := &kube_core.Node{}
+	if err := r.Get(ctx, req.NamespacedName, node); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			log.Error(err, "node not found")
+			return kube_ctrl.Result{}, nil
+		}
+		log.Error(err, "unable to fetch node")
+		return kube_ctrl.Result{}, err
+	}
+	log.Info("node successfully fetched")
+
+	kubeSystemPods := &kube_core.PodList{}
+	namespaceOption := kube_client.InNamespace(cniPodNamespace)
+	matchingFields := kube_client.MatchingFields{nodeIndexField: node.Name}
+	matchingLabels := kube_client.MatchingLabels{cniAppLabel: r.CniApp}
+	if err := r.Client.List(ctx, kubeSystemPods, namespaceOption, matchingFields, matchingLabels); err != nil {
+		return kube_ctrl.Result{}, err
+	}
+
+	err := r.updateTaints(ctx, log, node, kubeSystemPods.Items)
+	if err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "unable to update node taints")
+	}
+
+	return kube_ctrl.Result{}, err
+}
+
+func (r *CniNodeTaintReconciler) updateTaints(ctx context.Context, log logr.Logger, node *kube_core.Node, pods []kube_core.Pod) error {
+	taintIndex := slices.IndexFunc(node.Spec.Taints, func(taint kube_core.Taint) bool {
+		return taint.Key == nodeReadinessTaintKey && taint.Effect == kube_core.TaintEffectNoSchedule
+	})
+
+	if taintIndex >= 0 {
+		if r.hasCniPodRunning(log, pods) {
+			log.Info("has cni pod running and taint")
+			return r.untaintNode(ctx, log, node, taintIndex)
+		} else {
+			log.Info("has no cni pod running and taint")
+			return nil
+		}
+	} else {
+		if r.hasCniPodRunning(log, pods) {
+			log.V(1).Info("has cni pod running and no taint")
+			return nil
+		} else {
+			log.Info("has no cni pod running and no taint")
+			return r.taintNode(ctx, log, node)
+		}
+	}
+}
+
+func (r *CniNodeTaintReconciler) untaintNode(ctx context.Context, log logr.Logger, node *kube_core.Node, taintIndex int) error {
+	node.Spec.Taints = slices.Delete(node.Spec.Taints, taintIndex, taintIndex+1)
```

Comment:
Even this `slices` library has a weak API, I can't believe we still have to mess around with indexes. Why isn't there `slices.Filter`? :confounded:

Reply:
"Golang considered harmful"